agent/cache/cache.go

```diff
@@ -609,13 +604,91 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 	// identical calls to fetch will return the same waiter rather than
 	// perform multiple fetches.
 	entry.Fetching = true
+	entry.HasGoroutine = true
 	c.entries[key] = entry
 	metrics.SetGauge([]string{"consul", "cache", "entries_count"}, float32(len(c.entries)))
 	metrics.SetGauge([]string{"cache", "entries_count"}, float32(len(c.entries)))
 
-	tEntry := r.TypeEntry
 	// The actual Fetch must be performed in a goroutine.
-	go func() {
+	go c.launchBackgroundFetcher(key, r, entry)
+
+	return entry.Waiter
+}
+
+func (c *Cache) launchBackgroundFetcher(key string, r getOptions, initialEntry cacheEntry) {
+	defer c.mutateExistingEntry(key, r, func(entry *cacheEntry) {
+		if entry.HasGoroutine {
+			entry.HasGoroutine = false
+			entry.Fetching = false
+		}
+	})
+
+	var (
+		attempt uint
+		entry   = initialEntry
+	)
+	for {
+		if !c.runBackgroundFetcherOnce(key, r, entry, &attempt) {
+			return
+		}
+
+		var totalWait time.Duration
+
+		// If we're over the attempt minimum, start an exponential backoff.
+		if wait := backOffWait(attempt); wait > 0 {
+			totalWait += wait
+		}
+
+		// If we have a timer, wait for it
+		if r.TypeEntry.Opts.RefreshTimer > 0 {
+			totalWait += r.TypeEntry.Opts.RefreshTimer
+		}
+
+		select {
+		case <-time.After(totalWait):
+		case <-c.stopCh:
+			return // Check if cache was stopped
+		}
+
+		// Trigger. The "allowNew" field is false because in the time we were
+		// waiting to refresh we may have expired and got evicted. If that
+		// happened, we don't want to create a new entry.
+		r.Info.MustRevalidate = false
+		r.Info.MinIndex = 0
+
+		// We acquire a write lock because we may have to set Fetching to true.
+		c.entriesLock.Lock()
+
+		var ok bool
+		ok, _, entry = c.getEntryLocked(r.TypeEntry, key, r.Info)
+		if !ok || entry.Fetching {
```

Comment:
Locally I just thought of changing this line to `if !ok || !entry.HasGoroutine {` since the latter half of this conditional post-refactor seemed like dead code.

That seems to have triggered at least one test to occasionally double-close the waiter channel and panic.